EKF/gps_yaw_fusion.cpp

```diff
@@ -0,0 +1,433 @@
+
+	float R_YAW = 1.0f;
+	float predicted_hdg;
+	float H_YAW[4];
+	float measured_hdg;
+
+	// check if data has been set to NAN indicating no measurement
+	if (ISFINITE(_gps_sample_delayed.yaw)) {
+		// calculate the observed yaw angle of antenna array, converting a from body to antenna yaw measurement
+		measured_hdg = _gps_sample_delayed.yaw + _gps_yaw_offset;
+
+		// define the predicted antenna array vector and rotate into earth frame
+		Vector3f ant_vec_bf = {cosf(_gps_yaw_offset), sinf(_gps_yaw_offset), 0.0f};
+		Vector3f ant_vec_ef = _R_to_earth * ant_vec_bf;
+
+		// check if antenna array vector is within 30 degrees of vertical and therefore unable to provide a reliable heading
+		if (fabsf(ant_vec_ef(2)) > cosf(math::radians(30.0f)))  {
+			return;
+		}
+
+		// calculate predicted antenna yaw angle
+		predicted_hdg =  atan2f(ant_vec_ef(1),ant_vec_ef(0));
+
+		// calculate observation jacobian
+		float t2 = sinf(_gps_yaw_offset);
+		float t3 = cosf(_gps_yaw_offset);
+		float t4 = q0*q3*2.0f;
+		float t5 = q0*q0;
+		float t6 = q1*q1;
+		float t7 = q2*q2;
+		float t8 = q3*q3;
+		float t9 = q1*q2*2.0f;
+		float t10 = t5+t6-t7-t8;
+		float t11 = t3*t10;
+		float t12 = t4+t9;
+		float t13 = t3*t12;
+		float t14 = t5-t6+t7-t8;
+		float t15 = t2*t14;
+		float t16 = t13+t15;
+		float t17 = t4-t9;
+		float t19 = t2*t17;
+		float t20 = t11-t19;
+		float t18 = (t20*t20);
+		if (t18 < 1e-6f) {
+			return;
+		}
+		t18 = 1.0f / t18;
+		float t21 = t16*t16;
+		float t22 = sq(t11-t19);
+		if (t22 < 1e-6f) {
+			return;
+		}
+		t22 = 1.0f/t22;
+		float t23 = q1*t3*2.0f;
+		float t24 = q2*t2*2.0f;
+		float t25 = t23+t24;
+		float t26 = 1.0f/t20;
+		float t27 = q1*t2*2.0f;
+		float t28 = t21*t22;
+		float t29 = t28+1.0f;
+		if (fabsf(t29) < 1e-6f) {
+			return;
+		}
+		float t30 = 1.0f/t29;
+		float t31 = q0*t3*2.0f;
+		float t32 = t31-q3*t2*2.0f;
+		float t33 = q3*t3*2.0f;
+		float t34 = q0*t2*2.0f;
+		float t35 = t33+t34;
+
+		H_YAW[0] = (t35/(t11-t2*(t4-q1*q2*2.0f))-t16*t18*t32)/(t18*t21+1.0f);
+		H_YAW[1] = -t30*(t26*(t27-q2*t3*2.0f)+t16*t22*t25);
+		H_YAW[2] = t30*(t25*t26-t16*t22*(t27-q2*t3*2.0f));
+		H_YAW[3] = t30*(t26*t32+t16*t22*t35);
+
+		// using magnetic heading tuning parameter
+		R_YAW = sq(fmaxf(_params.mag_heading_noise, 1.0e-2f));
+
+	} else {
+		// there is nothing to fuse
+		return;
+	}
+
+	// wrap the heading to the interval between +-pi
+	measured_hdg = wrap_pi(measured_hdg);
+
+	// calculate the innovation and define the innovaton gate
+	float innov_gate = math::max(_params.heading_innov_gate, 1.0f);
+	_heading_innov = predicted_hdg - measured_hdg;
+
+	// wrap the innovation to the interval between +-pi
+	_heading_innov = wrap_pi(_heading_innov);
+
+	// Calculate innovation variance and Kalman gains, taking advantage of the fact that only the first 3 elements in H are non zero
+	// calculate the innovaton variance
+	float PH[4];
+	_heading_innov_var = R_YAW;
+
+	for (unsigned row = 0; row <= 3; row++) {
```

Comment:
its a hang over from early optimisation tests that showed slightly reduced execution times if the inner loop index type was unsigned instead of uint8_t.